pastas_plugins/pest/solver.py

```diff
@@ -1274,3 +1300,323 @@ def solve() -> None:
+        return data
+
+    def initialize(
+        self,
+        standard_deviation: float = 0.0,
+        correlation_coefficient: float = 0.0,
+        par_sigma_range: float = 4.0,
+        method: Literal["norm", "truncnorm", "uniform"] = "norm",
+    ) -> None:
+        logger.debug("Initialize: RML solver")
+        logger.debug("Initialize: Creating observation noise")
+        observations = self.ml.observations()
+        observation_noise = pd.DataFrame(
+            PestIesSolver.generate_observation_noise(
+                ies_num_reals=self.num_reals,
+                nobs=len(observations),
+                standard_deviation=standard_deviation,
+                correlation_coefficient=correlation_coefficient,
+                seed=self.seed,
+            ),
+            columns=pd.Index(range(self.num_reals)),
+            index=observations.index,
+        )
+
+        logger.debug("Initialize: Creating parameter noise")
+        parameter_data = pd.DataFrame(
+            index=pd.Index(range(self.num_reals)),
+            columns=self.ml.parameters.index,
+            dtype=float,
+        )
+        for pname, pdata in self.ml.parameters.iterrows():
+            rvs = PestIesSolver.parameter_distribution(
+                ies_num_reals=self.num_reals,
+                initial=pdata.at["initial"],
+                pmin=pdata.at["pmin"],
+                pmax=pdata.at["pmax"],
+                par_sigma_range=par_sigma_range,
+                method=method,
+            )
+            parameter_data[pname] = rvs
+        parameter_data.loc[:, :] = np.random.default_rng(seed=self.seed).permuted(
+            parameter_data.values, axis=0
+        )
+        if self.add_base:
+            logger.debug("Initialize: Adding base realization")
+            base_idx = self.num_reals - 1
+            observation_noise.loc[:, base_idx] = 0.0
+            observation_noise = observation_noise.rename(columns={base_idx: "base"})
+
+            parameter_data.loc[base_idx, :] = self.ml.parameters.loc[
+                :, "initial"
+            ].values
+            parameter_data = parameter_data.rename(index={base_idx: "base"})
+
+        self.parameter_ensemble = parameter_data
+        self.observation_noise = observation_noise
+
+    @property
+    def observation_ensemble(self) -> pd.DataFrame:
+        """Generate the observation ensemble by adding noise to model observations."""
+        obs = self.ml.observations().values
+        noise = self.observation_noise.values.T
+
+        obs_df = pd.DataFrame(
+            noise + obs,
+            index=self.observation_noise.columns,
+            columns=self.ml.observations().index,
+        ).T
+        return obs_df
+
+    @staticmethod
+    def jacobian_empirical(
+        simulation_ensembles: ArrayLike, parameter_ensembles: ArrayLike
+    ) -> ArrayLike:
+        """Approximate the Jacobian matrix using ensemble perturbations."""
+        return PestIesSolver.jacobian_empirical(
```

Comment:
Maybe move to a jacobian module? All these staticmethods hanging on to classes is nice, but becomes a bit chaotic when there are too many?

Reply:
see other comment